python/ql/test/experimental/dataflow/coverage/validTest.py

```diff
@@ -1,8 +1,8 @@
-def check_output(s, f):
-    if s == "OK\n":
+def check_output(outtext, f):
+    if all(s == "OK" for s in outtext.splitlines()):
         pass
     else:
-        raise RuntimeError("Function failed", s, f)
+        raise RuntimeError("Function failed", outtext, f)
```

Comment:
This looks like it will accept multiple lines of "OK". Have you checked that this does not mask any tests where only a single "OK" is expected? Did you add a test that needs this change?

Reply:
>This looks like it will accept multiple lines of "OK".

Yep (I also stated this in the commit message of https://github.com/github/codeql/pull/4159/commits/45ab723423fa4859bc0caf8ab2c3a138eb6a3b16)

> Have you checked that this does not mask any tests where only a single "OK" is expected?

I'm not sure what that means? But running the code worked fine :+1:

> Did you add a test that needs this change?

yes, the `test_swap` part (https://github.com/github/codeql/pull/4159/commits/45ab723423fa4859bc0caf8ab2c3a138eb6a3b16#diff-601e823041b1218704b390c8387e5bb6R371-R379)

Reply:
> I'm not sure what that means? But running the code worked fine 👍
My fear is that errors can hide in the imprecision. That when writing a test one expects a single "OK", but the interpreter actually calls the function multiple times or calls another function unexpectedly. If we do not run the function but simply run `isValid`, we would not see this. In `classes.py`, `OK` is not printed from `SINK` but rather from the expected special methods.
I guess in general, methods that should not be called should be made to print `NO` (or something similar != `OK`) and the multiple calls is a very specific case that would need careful testing anyway. So probably, we are fine as long as we update the comments about expected output.

Reply:
> interpreter actually calls the function multiple times

yes, that becomes a problem now. We _could_ change validTest so you need to decorate a test-function with `@multiple_ok`. I would rather make it your call if we should do this or not :wink: 

> or calls another function unexpectedly.

if the unexpected function prints OK, that wouldn't be caught by the current test-setup either :neutral_face:

Reply:
> if the unexpected function prints OK, that wouldn't be caught by the current test-setup either 😐
It would if the expected one is also called...

The decorator idea is nice, could that be made to specify the number of `OK`s as well?

I think for now, it is fine to just update the comments at the top of the test files referring to `validTest.py`.